testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/AbstractKeycloakTest.java

```diff
@@ -16,640 +16,16 @@
-import org.keycloak.common.util.Time;
-import org.keycloak.representations.idm.ClientRepresentation;
-import org.keycloak.representations.idm.RealmRepresentation;
-import org.keycloak.representations.idm.RequiredActionProviderRepresentation;
-import org.keycloak.representations.idm.UserRepresentation;
-import org.keycloak.services.resources.account.AccountFormService;
-import org.keycloak.testsuite.admin.ApiUtil;
-import org.keycloak.testsuite.arquillian.AuthServerTestEnricher;
-import org.keycloak.testsuite.arquillian.KcArquillian;
-import org.keycloak.testsuite.arquillian.SuiteContext;
-import org.keycloak.testsuite.arquillian.TestContext;
-import org.keycloak.testsuite.auth.page.AuthRealm;
-import org.keycloak.testsuite.auth.page.AuthServer;
-import org.keycloak.testsuite.auth.page.AuthServerContextRoot;
-import org.keycloak.testsuite.auth.page.WelcomePage;
-import org.keycloak.testsuite.auth.page.account.Account;
-import org.keycloak.testsuite.auth.page.login.OIDCLogin;
-import org.keycloak.testsuite.auth.page.login.UpdatePassword;
-import org.keycloak.testsuite.client.KeycloakTestingClient;
-import org.keycloak.testsuite.pages.LoginPasswordUpdatePage;
-import org.keycloak.testsuite.util.DroneUtils;
-import org.keycloak.testsuite.util.OAuthClient;
-import org.keycloak.testsuite.util.TestCleanup;
-import org.keycloak.testsuite.util.TestEventsLogger;
-import org.openqa.selenium.WebDriver;
-
-import javax.ws.rs.NotFoundException;
-import javax.ws.rs.core.UriBuilder;
-
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.PipedInputStream;
-import java.io.PipedOutputStream;
-import java.io.PrintWriter;
-import java.net.URI;
-import java.net.URISyntaxException;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Calendar;
-import java.util.Collections;
-import java.util.List;
-import java.util.Map;
-import java.util.Scanner;
-
-import static org.hamcrest.Matchers.equalTo;
-import static org.hamcrest.Matchers.is;
-import static org.junit.Assert.assertThat;
-import static org.keycloak.testsuite.admin.Users.setPasswordFor;
-import static org.keycloak.testsuite.util.ServerURLs.AUTH_SERVER_HOST;
-import static org.keycloak.testsuite.util.ServerURLs.AUTH_SERVER_PORT;
-import static org.keycloak.testsuite.util.ServerURLs.AUTH_SERVER_SCHEME;
-import static org.keycloak.testsuite.util.ServerURLs.AUTH_SERVER_SSL_REQUIRED;
-import static org.keycloak.testsuite.auth.page.AuthRealm.MASTER;
-import static org.keycloak.testsuite.util.URLUtils.navigateToUri;
-import static org.keycloak.testsuite.util.ServerURLs.removeDefaultPorts;
+import org.keycloak.common.Profile;
+import org.keycloak.testsuite.arquillian.annotation.DisableFeature;
 
 /**
- *
- * @author tkyjovsk
+ * Old tests that use AbstractKeycloakTest assume that we are using the old
+ * account console.
+ * 
+ * @author Stan Silvert
  */
-@RunWith(KcArquillian.class)
-@RunAsClient
-public abstract class AbstractKeycloakTest {
```

Comment:
I have to admit I don't like the approach with renaming and creating new abstract classes. The naming is quite confusing – `AbstractKeycloakTest` vs `MainAbstractKeycloakTest`. But more importantly, it breaks the class hierarchy which results in duplicate code e.g. in `AbstractUiTest`.

The way I see it we have two options.
1. Move the abstract class that disables the new console much, much lower in the hierarchy. So that the abstract classes could be used by the new console tests and code duplication wouldn't be necessary.
2. Disable the new console in the first abstract class and then re-enable it only in the tests that need it. The number of tests for the new console is really just a fraction of what we have, so maybe that would be a cleaner approach.

In any case, this change should be as small as possible because it's only temporary. Sooner or later the old console will be deprecated and we'll have to fix all the tests that rely on it.

Reply:
> I have to admit I don't like the approach with renaming and creating new abstract classes. The naming is quite confusing – `AbstractKeycloakTest` vs `MainAbstractKeycloakTest`. But more importantly, it breaks the class hierarchy which results in duplicate code e.g. in `AbstractUiTest`.
> 
> The way I see it we have two options.
> 
> 1. Move the abstract class that disables the new console much, much lower in the hierarchy. So that the abstract classes could be used by the new console tests and code duplication wouldn't be necessary.
> 2. Disable the new console in the first abstract class and then re-enable it only in the tests that need it. The number of tests for the new console is really just a fraction of what we have, so maybe that would be a cleaner approach.
> 
> In any case, this change should be as small as possible because it's only temporary. Sooner or later the old console will be deprecated and we'll have to fix all the tests that rely on it.

I'm not really happy with this either.

I spent a fair amount of time trying to figure out what would be the least disruptive to the current code base.

I don't think your first suggestion would work at all without changing a lot of tests.  If I recall, a lot of tests extend AbstractKeycloakTest directly.

I wouldn't want to go through all the tests that need new account console and enable NAC individually.  That just seems wrong since we want to use NAC going forward.

IMO, the only thing that made the code a little more ugly was that I had to copy some code from AbstractAuthorizationTest to AbstractUITest.  Because of the way the hierarchy is set up there was no good way to get around it without figuring out exactly what each test was using from the base class.

This is what happens when a code base relies so heavily on inheritance instead of delegation.

That being said, the whole thing could definitely use some refactoring but it's just a matter of deciding how much time you want to spend.

Reply:
> I don't think your first suggestion would work at all without changing a lot of tests. If I recall, a lot of tests extend AbstractKeycloakTest directly.

Ah, yes, you're correct. Option 1 won't work. 
> I wouldn't want to go through all the tests that need new account console and enable NAC individually. That just seems wrong since we want to use NAC going forward.

Actually, I think you wouldn't need to go trough all the tests that need NAC. The annotations to enable it are already there, the only change we'd have to do now is to disable NAC in the most abstract class. Once we fixed the tests to work with NAC, we'll just remove those annotations. That seems to me like the least invasive option. If it works, of course. ;)

> That being said, the whole thing could definitely use some refactoring but it's just a matter of deciding how much time you want to spend.

+1000
That being said, it's ugly but we don't need to make it even more ugly now. :)

Reply:
> Actually, I think you wouldn't need to go trough all the tests that need NAC. The annotations to enable it are already there, the only change we'd have to do now is to disable NAC in the most abstract class. Once we fixed the tests to work with NAC, we'll just remove those annotations. That seems to me like the least invasive option. If it works, of course. ;)
> 
I tried that and it didn't appear to work.  Even if it did things would get confusing.  You end up with an inheritance hierarchy where a base class has `@EnableFeature` and a subclass has `@DisableFeature`.  Or vice versa.  So you don't know who wins.

I did find that using the annotation on a method takes precedence, over having it at the class level.  So that helped out a bit.

Reply:
IMHO there's no problem with a parent class disabling feature and some child re-enabling it. I'm sure we could fix the implementation if it doesn't work like that.

As an alternative, we could globally disable NAS on testsuite level by some other way (without using the annotation) and use the annotation just for enabling where needed. It's all just temporary anyway so we should do as little changes in the class hierarchy as possible.

Reply:
> IMHO there's no problem with a parent class disabling feature and some child re-enabling it. I'm sure we could fix the implementation if it doesn't work like that.
> 
I think it is a problem because there is no hard and fast rule about who should win.  And even if you know how to determine the winner you must inspect the entire hierarchy to find all the @Enable and @Disable annotations.  That's very confusing for the developer.

> As an alternative, we could globally disable NAS on testsuite level by some other way (without using the annotation) and use the annotation just for enabling where needed. It's all just temporary anyway so we should do as little changes in the class hierarchy as possible.

We both agree that things have gotten messy.  Let's fix it in another JIRA.

Reply:
> I think it is a problem because there is no hard and fast rule about who should win. And even if you know how to determine the winner you must inspect the entire hierarchy to find all the @enable and @disable annotations. That's very confusing for the developer.

It already works like that – the entire hierarchy is transparently checked for the annotations thanks to [`@Inherited`](https://github.com/keycloak/keycloak/blob/74988a3f21c26764e01956ff8c9ef5cea865cb2d/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/annotation/EnableFeature.java#L19). ;) And to clearly determine who wins is just a matter of updating [`checkAnnotatedElementForFeatureAnnotations`](https://github.com/keycloak/keycloak/blob/74988a3f21c26764e01956ff8c9ef5cea865cb2d/testsuite/integration-arquillian/tests/base/src/main/java/org/keycloak/testsuite/arquillian/containers/KeycloakContainerFeaturesController.java#L148-L168). I still believe this is a cleaner way than to break the class hierarchy.

> We both agree that things have gotten messy. Let's fix it in another JIRA.

Sure, a follow-up JIRA can fix things but IMHO we don't need to make it even more messy in this PR. :)

Reply:
@ssilvert PTAL at https://github.com/ssilvert/keycloak/pull/8. It makes the enabling/disabling features more predictable so that we can disable NAC in abstract test class and the re-enable it in NAC tests.

@mhajas Could you please check https://github.com/ssilvert/keycloak/pull/8 if it makes sense to you? It touches your `KeycloakContainerFeaturesController`.

Reply:
@mhajas Sure, I plan to do that once @ssilvert approves the changes as well. Thank you for checking it!